tests/unit/testplan/testing/test_base.py

```diff
@@ -0,0 +1,108 @@
+"""TODO."""
+
+import time
+
+from testplan import TestplanMock
+from testplan.common.entity import Resource, Runnable
+from testplan.testing.base import Test, TestResult
+from testplan.runnable import TestRunnerStatus
+from testplan.runners.local import LocalRunner
+from testplan.report import Status, TestGroupReport, ReportCategories
+
+
+class DummyDriver(Resource):
+    def start(self):
+        self.status.change(self.STATUS.STARTING)
+        self.starting()
+
+    def stop(self):
+        if self.status.tag is self.STATUS.STOPPED:
+            return
+        self.status.change(self.STATUS.STOPPING)
+        if self.active:
+            self.stopping()
+
+    def starting(self):
+        time.sleep(0.2)  # 200ms for startup
+
+    def stopping(self):
+        time.sleep(0.1)  # 100ms for teardown
+
+    def aborting(self):
+        pass
+
+
+class DummyTest(Test):
+    def __init__(self, name, **options):
+        super(DummyTest, self).__init__(name=name, **options)
+        self.resources.add(DummyDriver(), uid="drv1")
+        self.resources.add(DummyDriver(), uid="drv2")
+
+    def run_tests(self):
+        with self._result.report.timer.record("run"):
+            time.sleep(0.5)  # 500ms for execution
+            self._result.report.status_override = Status.PASSED
+
+    def pre_resource_steps(self):
+        self._add_step(lambda: self.result.report.timer.start("setup"))
```

Comment:
I think this logic is already in Test class, so DummyTest shall just use the inherited pre_resource_steps and we can test that time is noted.

Reply:
updated. remove "setup" and "teardown" field and add other flags for testing.